src/tools/wasm-link.cc

```diff
@@ -555,14 +553,20 @@ static void write_reloc_section(Context* ctx,
   write_u32_leb128_enum(&ctx->stream, section_code, "reloc section");
   write_u32_leb128(&ctx->stream, total_relocs, "num relocs");
 
-  for (size_t i = 0; i < sections.size(); i++) {
-    Section* sec = sections[i];
-    const std::vector<Reloc>& relocs = sec->relocations;
-    for (size_t j = 0; j < relocs.size(); j++) {
-      write_u32_leb128_enum(&ctx->stream, relocs[j].type, "reloc type");
-      uint32_t new_offset = relocs[j].offset + sec->output_payload_offset;
+  for (Section* sec: sections) {
+    for (const Reloc& reloc: sec->relocations) {
+      write_u32_leb128_enum(&ctx->stream, reloc.type, "reloc type");
+      uint32_t new_offset = reloc.offset + sec->output_payload_offset;
       write_u32_leb128(&ctx->stream, new_offset, "reloc offset");
-      write_u32_leb128(&ctx->stream, relocs[j].index, "reloc index");
+      uint32_t index = reloc.index;
+      switch (reloc.type) {
+        case RelocType::FuncIndexLEB:
+          index = relocate_func_index(sec->binary, index);
+          break;
+        default:
+          break;
```

Comment:
Should this switch statement be expanded to cover, say, table indices, and possibly global indices? I think I'd prefer a fatal error as the `default`, and an explicit list of reloc types that don't need relocating before the `break`.